builtin/providers/openstack/resource_openstack_networking_router_route_v2.go

```diff
@@ -0,0 +1,216 @@
+		}
+	} else {
+		log.Printf("[DEBUG] Router %s has route already", routerId)
+	}
+
+	return resourceNetworkingRouterRouteV2Read(d, meta)
+}
+
+func resourceNetworkingRouterRouteV2Read(d *schema.ResourceData, meta interface{}) error {
+
+	routerId := d.Get("router_id").(string)
+
+	config := meta.(*Config)
+	networkingClient, err := config.networkingV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack networking client: %s", err)
+	}
+
+	n, err := routers.Get(networkingClient, routerId).Extract()
+	if err != nil {
+		httpError, ok := err.(*gophercloud.UnexpectedResponseCodeError)
+		if !ok {
+			return fmt.Errorf("Error retrieving OpenStack Neutron Router: %s", err)
+		}
+
+		if httpError.Actual == 404 {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error retrieving OpenStack Neutron Router: %s", err)
+	}
+
+	log.Printf("[DEBUG] Retrieved Router %s: %+v", routerId, n)
+
+	var destCidr string = d.Get("destination_cidr").(string)
+	var nextHop string = d.Get("next_hop").(string)
+
+	d.Set("next_hop", "")
+	d.Set("destination_cidr", "")
+
+	for _, r := range n.Routes {
+
+		if r.DestinationCIDR == destCidr && r.NextHop == nextHop {
+			d.Set("destination_cidr", destCidr)
+			d.Set("next_hop", nextHop)
+			break
+		}
+	}
+
+	return nil
+}
+
+func resourceNetworkingRouterRouteV2Delete(d *schema.ResourceData, meta interface{}) error {
+
+	routerId := d.Get("router_id").(string)
+	osMutexKV.Lock(routerId)
+	defer osMutexKV.Unlock(routerId)
+
+	config := meta.(*Config)
+
+	networkingClient, err := config.networkingV2Client(d.Get("region").(string))
+	if err != nil {
+		return fmt.Errorf("Error creating OpenStack networking client: %s", err)
+	}
+
+	n, err := routers.Get(networkingClient, routerId).Extract()
+	if err != nil {
+		httpError, ok := err.(*gophercloud.UnexpectedResponseCodeError)
+		if !ok {
+			return fmt.Errorf("Error retrieving OpenStack Neutron Router: %s", err)
+		}
+
+		if httpError.Actual == 404 {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error retrieving OpenStack Neutron Router: %s", err)
+	}
+
+	var updateOpts routers.UpdateOpts
+
+	var destCidr string = d.Get("destination_cidr").(string)
+	var nextHop string = d.Get("next_hop").(string)
+
+	var oldRts []routers.Route = n.Routes
+	var newRts []routers.Route
+
+	for _, r := range oldRts {
+
+		if r.DestinationCIDR != destCidr || r.NextHop != nextHop {
+			newRts = append(newRts, r)
+		}
+	}
+
+	if len(oldRts) != len(newRts) {
+		r := routers.Route{DestinationCIDR: destCidr, NextHop: nextHop}
+		log.Printf(
+			"[INFO] Deleting route %s", r)
+		updateOpts.Routes = newRts
+
+		log.Printf("[DEBUG] Updating Router %s with options: %+v", routerId, updateOpts)
+
+		_, err = routers.Update(networkingClient, routerId, updateOpts).Extract()
+		if err != nil {
+			return fmt.Errorf("Error updating OpenStack Neutron Router: %s", err)
+		}
+	} else {
+		return fmt.Errorf("Route did not exist already")
+	}
+
+	d.SetId("")
```

Comment:
I recently learned that it's no longer required to `SetId("")` any more. A successful return of the Delete function will do this automatically. A lot of resources still have it laying around, but newer ones are/can omitting it.